### PR TITLE
{Packaging} Deprecation notice urllib3[secure]

### PR DESCRIPTION
**Pull request by fyunusa**
### Description

pyOpenSSL and urllib3[secure] are deprecated in the upcoming release (1.26.12)
https://github.com/urllib3/urllib3/issues/2680

### CODE REVIEW

**Comment by yonzhan** [comment]
appservice

File: src/azure-cli/setup.py
Comment (sethmlarson) [inline src/azure-cli/setup.py:152]: You may want to retain the dependencies that were originally installed via `urllib3[secure]` here like `idna, pyOpenSSL, etc. That it up for the Azure CLI team to decide.

**Comment by zhoxing-ms** [comment]
@panchagnula Could you please help review this PR?

File: src/azure-cli/setup.py
Comment (zhoxing-ms) [reply src/azure-cli/setup.py:152]: May I ask why you want to modify this?

**Comment by zhoxing-ms** [comment]
> @[zhoxing-ms](https://github.com/zhoxing-ms) zhoxing-ms deleted a comment from [sethmlarson](https://github.com/sethmlarson) [21 seconds ago](https://github.com/Azure/azure-cli/pull/23494#event-7286784836)

Sorry, I only wanted to delete my own comment, but I deleted the whole comment by mistake

**Comment by panchagnula** [comment]
@StrawnSC could you also check if we don't use any of these libraries in ACA extension & might need to upgrade as well. I can do the testing for this PR/ webapp

**Comment by StrawnSC** [comment]
> @StrawnSC could you also check if we don't use any of these libraries in ACA extension & might need to upgrade as well. I can do the testing for this PR/ webapp

@panchagnula I just checked and we don't use this library in the ACA extension

**Comment by doubleyewdee** [comment]
Bumping this to bring it back to attention. @zhoxing-ms any chance we can pull this in (assuming a subsequent merge/CI test passes)?

Asking because we have some code over in Azure ML that uses the CLI and is seeing deprecation warnings here. Thanks!

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:2596]: This code was introduced by https://github.com/Azure/azure-cli/pull/907 in 2016. According to https://github.com/urllib3/urllib3/issues/2680

> If you're using Python 2.7.10 or later and OpenSSL 1.1.1+ (this is almost everyone) then you won't need the urllib3[secure] extra or urllib3.contrib.pyopenssl module. Both were necessary a long time ago, but aren't necessary any more!

this line is no longer needed since we have dropped support for Python 2 (https://github.com/Azure/azure-cli/pull/11363) and most distributions now have OpenSSL 1.1.1+ or even 3.0.

In addition, instead of using `urllib3`, we should use `requests` as unified way to make HTTP requests (https://github.com/Azure/azure-cli/issues/26008).

**Comment by fyunusa** [comment]
Ok will do so shortly

**Comment by fyunusa** [comment]
> @fyunusa please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> 
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
>     1. **Definitions**.
>        **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>        that is delivered by You to Microsoft under this Agreement.
>        **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>        approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>        **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>        Project, including but not limited to communication on electronic mailing lists, source code control
>        systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>        discussing and improving that Project, but excluding communication that is conspicuously marked or
>        otherwise designated in writing by You as “Not a Submission.”
>        **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>        associated comments and documentation.
> 
>     2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>        Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>        described in Section 4 below), Submit to any Project.
> 
>     3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>        Should You wish to Submit materials that are not Your original work, You may Submit them separately
>        to the Project if You (a) retain all copyright and license information that was in the materials as You
>        received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>        containing materials of a third party:” followed by the names of the third party and any licenses or other
>        restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>        guidelines concerning Submissions.
> 
>     4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>        for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>        Submission is made in the course of Your work for an employer or Your employer has intellectual
>        property rights in Your Submission by contract or applicable law, You must secure permission from Your
>        employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>        Agreement will refer to You and the employer collectively. If You change employers in the future and
>        desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>        and secure permission from the new employer before Submitting those Submissions.
> 
>     5. **Licenses**.
> 
> 
>     * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>       indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>       Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>       the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>       parties.
> 
>     * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>       indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>       Your patent claims that are necessarily infringed by the Submission or the combination of the
>       Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>       import or otherwise dispose of the Submission alone or with the Project.
> 
>     * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>       No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>       granted by implication, exhaustion, estoppel or otherwise.
> 
> 
>     6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>        licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>        have disclosed under Section 3). You represent that You have secured permission from Your employer to
>        make the Submission in cases where Your Submission is made in the course of Your work for Your
>        employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>        law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>        have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>        You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>        REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>        EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>        PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>        NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 
>     7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>        You later become aware that would make Your representations in this Agreement inaccurate in any
>        respect.
> 
>     8. **Information about Submissions**. You agree that contributions to Projects and information about
>        contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>        information that You submit with Your Submission.
> 
>     9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>        the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>        Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>        exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>        defenses of lack of personal jurisdiction and forum non-conveniens.
> 
>     10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>         supersedes any and all prior agreements, understandings or communications, written or oral, between
>         the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

@microsoft-github-policy-service agree

**Comment by jiasli** [comment]
@fyunusa, Thanks for the PR. However, as there are many linter errors:

https://dev.azure.com/azclitools/public/_build/results?buildId=58123&view=logs&jobId=36dd4138-4d53-5e46-00d9-e5cd9744be05&j=36dd4138-4d53-5e46-00d9-e5cd9744be05&t=1cf3879c-0a42-5ccd-7693-7a4781d739d8

```
ERROR: ************* Module azure.cli.command_modules.appservice.custom
src/azure-cli/azure/cli/command_modules/appservice/custom.py:2798:0: C0303: Trailing whitespace (trailing-whitespace)
src/azure-cli/azure/cli/command_modules/appservice/custom.py:2810:0: C0303: Trailing whitespace (trailing-whitespace)
src/azure-cli/azure/cli/command_modules/appservice/custom.py:2800:8: W0621: Redefining name 'ssl' from outer scope (line 19) (redefined-outer-name)
src/azure-cli/azure/cli/command_modules/appservice/custom.py:2800:8: W0404: Reimport 'ssl' (imported line 19) (reimported)
src/azure-cli/azure/cli/command_modules/appservice/custom.py:2797:4: W0611: Unused import urllib3 (unused-import)
************* Module azure.cli.command_modules.appservice.tunnel
src/azure-cli/azure/cli/command_modules/appservice/tunnel.py:80:12: W0621: Redefining name 'ssl' from outer scope (line 8) (redefined-outer-name)
src/azure-cli/azure/cli/command_modules/appservice/tunnel.py:80:12: W0404: Reimport 'ssl' (imported line 8) (reimported)
src/azure-cli/azure/cli/command_modules/appservice/custom.py:2798:0: C0303: Trailing whitespace (trailing-whitespace)
src/azure-cli/azure/cli/command_modules/appservice/custom.py:2810:0: C0303: Trailing whitespace (trailing-whitespace)
src/azure-cli/azure/cli/command_modules/appservice/custom.py:2800:8: W0621: Redefining name 'ssl' from outer scope (line 19) (redefined-outer-name)
src/azure-cli/azure/cli/command_modules/appservice/custom.py:2800:8: W0404: Reimport 'ssl' (imported line 19) (reimported)
src/azure-cli/azure/cli/command_modules/appservice/custom.py:2797:4: W0611: Unused import urllib3 (unused-import)
src/azure-cli/azure/cli/command_modules/appservice/tunnel.py:80:12: W0621: Redefining name 'ssl' from outer scope (line 8) (redefined-outer-name)
src/azure-cli/azure/cli/command_modules/appservice/tunnel.py:80:12: W0404: Reimport 'ssl' (imported line 8) (reimported)
```

Could you rebase to the latest `dev` branch and only keep the change for `src/azure-cli/setup.py`. I will merge your PR so you can get credit.

I will then deal with `src/azure-cli/azure/cli/command_modules/appservice/custom.py` in a separate PR.

File: src/azure-cli/setup.py
Comment (jiasli) [inline src/azure-cli/setup.py:152]: `urllib3[secure]` installs below packages: 

https://github.com/urllib3/urllib3/blob/e67f13c44fa4aada1954df96eb8cccbb5f2dbd2a/pyproject.toml#L49-L55

```toml
secure = [
  "pyOpenSSL>=17.1.0",
  "cryptography>=1.9",
  "idna>=2.0.0",
  "certifi",
  "urllib3-secure-extra",
]
```

We do explicitly require them in our `requirements.txt`: https://github.com/Azure/azure-cli/blob/424374558bedf9031069e5a70e712dbe79a2442c/src/azure-cli/requirements.py3.Linux.txt

This complies with the migration guidance: https://github.com/urllib3/urllib3/issues/2700

>   * If needed, add `pyOpenSSL>=0.14`, `cryptography>=1.3.4`, `idna>=2`, and `certifi` to their dependencies. These dependencies should only be added back if they're actually used within the project. If they're not used (including `pyopenssl.inject_into_urllib3`) then they can likely be omitted.

Besides, according to https://urllib3.readthedocs.io/en/stable/reference/contrib/pyopenssl.html

> However, pyOpenSSL depends on cryptography, which depends on idna, so while we use all three directly here we end up having relatively few packages required.

**Comment by fyunusa** [comment]
> @fyunusa, Thanks for the PR. However, as there are many linter errors:
> 
> https://dev.azure.com/azclitools/public/_build/results?buildId=58123&view=logs&jobId=36dd4138-4d53-5e46-00d9-e5cd9744be05&j=36dd4138-4d53-5e46-00d9-e5cd9744be05&t=1cf3879c-0a42-5ccd-7693-7a4781d739d8
> 
> ```
> ERROR: ************* Module azure.cli.command_modules.appservice.custom
> src/azure-cli/azure/cli/command_modules/appservice/custom.py:2798:0: C0303: Trailing whitespace (trailing-whitespace)
> src/azure-cli/azure/cli/command_modules/appservice/custom.py:2810:0: C0303: Trailing whitespace (trailing-whitespace)
> src/azure-cli/azure/cli/command_modules/appservice/custom.py:2800:8: W0621: Redefining name 'ssl' from outer scope (line 19) (redefined-outer-name)
> src/azure-cli/azure/cli/command_modules/appservice/custom.py:2800:8: W0404: Reimport 'ssl' (imported line 19) (reimported)
> src/azure-cli/azure/cli/command_modules/appservice/custom.py:2797:4: W0611: Unused import urllib3 (unused-import)
> ************* Module azure.cli.command_modules.appservice.tunnel
> src/azure-cli/azure/cli/command_modules/appservice/tunnel.py:80:12: W0621: Redefining name 'ssl' from outer scope (line 8) (redefined-outer-name)
> src/azure-cli/azure/cli/command_modules/appservice/tunnel.py:80:12: W0404: Reimport 'ssl' (imported line 8) (reimported)
> src/azure-cli/azure/cli/command_modules/appservice/custom.py:2798:0: C0303: Trailing whitespace (trailing-whitespace)
> src/azure-cli/azure/cli/command_modules/appservice/custom.py:2810:0: C0303: Trailing whitespace (trailing-whitespace)
> src/azure-cli/azure/cli/command_modules/appservice/custom.py:2800:8: W0621: Redefining name 'ssl' from outer scope (line 19) (redefined-outer-name)
> src/azure-cli/azure/cli/command_modules/appservice/custom.py:2800:8: W0404: Reimport 'ssl' (imported line 19) (reimported)
> src/azure-cli/azure/cli/command_modules/appservice/custom.py:2797:4: W0611: Unused import urllib3 (unused-import)
> src/azure-cli/azure/cli/command_modules/appservice/tunnel.py:80:12: W0621: Redefining name 'ssl' from outer scope (line 8) (redefined-outer-name)
> src/azure-cli/azure/cli/command_modules/appservice/tunnel.py:80:12: W0404: Reimport 'ssl' (imported line 8) (reimported)
> ```
> 
> Could you rebase to the latest `dev` branch and only keep the change for `src/azure-cli/setup.py`. I will merge your PR so you can get credit.
> 
> I will then deal with `src/azure-cli/azure/cli/command_modules/appservice/custom.py` in a separate PR.
> 
> 

Ok then I'll work on that

**Comment by fyunusa** [comment]
> > @fyunusa, Thanks for the PR. However, as there are many linter errors:
> > https://dev.azure.com/azclitools/public/_build/results?buildId=58123&view=logs&jobId=36dd4138-4d53-5e46-00d9-e5cd9744be05&j=36dd4138-4d53-5e46-00d9-e5cd9744be05&t=1cf3879c-0a42-5ccd-7693-7a4781d739d8
> > ```
> > ERROR: ************* Module azure.cli.command_modules.appservice.custom
> > src/azure-cli/azure/cli/command_modules/appservice/custom.py:2798:0: C0303: Trailing whitespace (trailing-whitespace)
> > src/azure-cli/azure/cli/command_modules/appservice/custom.py:2810:0: C0303: Trailing whitespace (trailing-whitespace)
> > src/azure-cli/azure/cli/command_modules/appservice/custom.py:2800:8: W0621: Redefining name 'ssl' from outer scope (line 19) (redefined-outer-name)
> > src/azure-cli/azure/cli/command_modules/appservice/custom.py:2800:8: W0404: Reimport 'ssl' (imported line 19) (reimported)
> > src/azure-cli/azure/cli/command_modules/appservice/custom.py:2797:4: W0611: Unused import urllib3 (unused-import)
> > ************* Module azure.cli.command_modules.appservice.tunnel
> > src/azure-cli/azure/cli/command_modules/appservice/tunnel.py:80:12: W0621: Redefining name 'ssl' from outer scope (line 8) (redefined-outer-name)
> > src/azure-cli/azure/cli/command_modules/appservice/tunnel.py:80:12: W0404: Reimport 'ssl' (imported line 8) (reimported)
> > src/azure-cli/azure/cli/command_modules/appservice/custom.py:2798:0: C0303: Trailing whitespace (trailing-whitespace)
> > src/azure-cli/azure/cli/command_modules/appservice/custom.py:2810:0: C0303: Trailing whitespace (trailing-whitespace)
> > src/azure-cli/azure/cli/command_modules/appservice/custom.py:2800:8: W0621: Redefining name 'ssl' from outer scope (line 19) (redefined-outer-name)
> > src/azure-cli/azure/cli/command_modules/appservice/custom.py:2800:8: W0404: Reimport 'ssl' (imported line 19) (reimported)
> > src/azure-cli/azure/cli/command_modules/appservice/custom.py:2797:4: W0611: Unused import urllib3 (unused-import)
> > src/azure-cli/azure/cli/command_modules/appservice/tunnel.py:80:12: W0621: Redefining name 'ssl' from outer scope (line 8) (redefined-outer-name)
> > src/azure-cli/azure/cli/command_modules/appservice/tunnel.py:80:12: W0404: Reimport 'ssl' (imported line 8) (reimported)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Could you rebase to the latest `dev` branch and only keep the change for `src/azure-cli/setup.py`. I will merge your PR so you can get credit.
> > I will then deal with `src/azure-cli/azure/cli/command_modules/appservice/custom.py` in a separate PR.
> 
> Ok then I'll work on that

@jiasli I've sync my repo to get latest update from the dev branch and then re sync the pull request, does that work fine?

**Comment by jiasli** [comment]
@fyunusa, there are still linter errors. Could you revert `src/azure-cli/azure/cli/command_modules/appservice/custom.py` and `src/azure-cli/azure/cli/command_modules/appservice/tunnel.py` to their original versions?

I will handle them in https://github.com/Azure/azure-cli/pull/23494.

**Comment by fyunusa** [comment]
> @fyunusa, there are still linter errors. Could you revert `src/azure-cli/azure/cli/command_modules/appservice/custom.py` and `src/azure-cli/azure/cli/command_modules/appservice/tunnel.py` to their original versions?
> 
> I will handle them in https://github.com/Azure/azure-cli/pull/26538.
> 

Ok

**Comment by jiasli** [comment]
@fyunusa, could you please make the change to make the CI pass? Thanks a lot.

**Comment by fyunusa** [comment]
> @fyunusa, could you please make the change to let the CI pass? Thanks a lot.
> 

Ok, will work on it as soon

**Comment by fyunusa** [comment]
> @fyunusa, could you please make the change to let the CI pass? Thanks a lot.

You can check now, so i've updated the tunnel.py & custom.py file to match the version on the dev

**Comment by fyunusa** [comment]
@jiasli i think all checks passed now, so you can review the PR and let me know if my attention is needed

**Review by jiasli — APPROVED** [review]

**Comment by jiasli** [comment]
Thank you very much @fyunusa. There is nothing pending on your side now. App Service team will review this PR soon. I will merge it after they approve.

**Review by bebound — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**
